### PR TITLE
fix(cli): reuse and restore orchestrator sessions across ao start cycles

### PR DESCRIPTION
**Pull request by harsh-batheja**
## Summary

Fixes #1048 — `ao start` was always allocating a new `{prefix}-orchestrator-N` instead of reattaching to the previous one, and the dashboard was linking to a stale legacy id (`{projectId}-orchestrator`) that the CLI never used.

Three coordinated changes:

- **Reuse-or-restore in `runStartup`** — list the project's orchestrators, partition into **live** (`!isTerminalSession`) and **restorable** (`isTerminalSession && isRestorable`) buckets, pick the most-recently-active candidate across both. Live → reuse the id in place. Restorable → call `sm.restore()` so the original numbered id is preserved (no fresh `-N` allocated). Only spawn when both buckets are empty. Restored sessions get an explicit `(restored)` marker in the CLI summary so the resurfaced id isn't a surprise.
- **Drop the phantom `${prefix}-orchestrator` id** from `runStartup` and `registerStop`. Every URL print, browser-open target, and kill target now uses the real numbered id. `ao stop` resolves the actual orchestrator via `sm.list(projectId)` (with the same `isOrchestratorSession` filter the dashboard uses) and reports the killed id via `console.log` so non-TTY callers and tests can see it.
- **Tighten `isOrchestratorSession`** in `packages/core/src/types.ts`. The bare `endsWith("-orchestrator")` shape now only counts when `metadata.role === "orchestrator"` is explicitly stamped. Session-manager's `repairSingleSessionMetadataOnRead` already backfills `role: orchestrator` for legacy records loaded via `sm.list()`, so the only records that get rejected are truly abandoned bare ids — exactly the ones that were leaking into the dashboard's orchestrator list with the wrong prefix.

## Why the bug existed

`runStartup` already had a reuse path, but it relied on `sm.list` returning the previous orchestrator as non-terminal. Once the orchestrator's tmux pane was destroyed, `enrichSessionWithRuntimeState` mutated `status` to `killed`, the `!isTerminalSession` filter dropped it, and the CLI fell through to spawn a new id every time. Meanwhile `ao stop` was calling `sm.kill("${prefix}-orchestrator")` against a phantom (no-N) id that never matched a real numbered session, so the actual orchestrator stayed alive on disk while the CLI kept allocating new numbers (`-3`, `-4`, …).

The dashboard amplified the confusion: stale `{projectId}-orchestrator` records (no number, no role metadata, no runtime handle to liveness-check) survived `isOrchestratorSession` via the `endsWith("-orchestrator")` fallback and outranked the real numbered sessions in `listDashboardOrchestrators`. The dashboard's "orchestrator" link pointed at the legacy id while the CLI printed a numbered id — exactly the mismatch the issue reported.

## Tests

- `packages/core/src/__tests__/types.test.ts` — regression tests for `isOrchestratorSession` rejecting bare `{projectId}-orchestrator` ids without role metadata, and accepting them when role is stamped. Existing positive cases for numbered ids and explicit `role: "orchestrator"` still pass.
- `packages/cli/__tests__/commands/start.test.ts` —
  - 4 new orchestrator-selection tests: restore-on-killed, ignore-stale-bare, cross-bucket recency picking, and live-multi reuse with the `N other session(s) available` suffix.
  - Updated `ao stop` tests to mock `sm.list` (the new lookup path) and assert the real numbered id gets killed.
  - Re-prime web-dir mocks in `beforeEach` so `findFreePort`/`isPortAvailable` aren't wiped by `vi.restoreAllMocks()` between tests (was masking `http://localhost:undefined` failures).
- `packages/web/src/lib/__tests__/serialize.test.ts` — 3 new `listDashboardOrchestrators` tests: stale-bare exclusion, numbered-live inclusion, role-stamped legacy inclusion.
- `packages/web/src/__tests__/orchestrators.test.tsx` — updated existing test to use a numbered orchestrator id with explicit role metadata.

All three test suites pass:
- `pnpm --filter @composio/ao-cli test` → 315 passed
- `pnpm --filter @composio/ao-core test` → 580 passed
- `pnpm --filter @composio/ao-web test` → 553 passed

## Out of scope

- No archival of zombie orchestrators — risky and orthogonal, better as a follow-up.
- The numbering scheme (`-N`) is unchanged. Reuse always wins over allocation, so under normal use the number stops growing.
- The web `POST /api/orchestrators` spawn path is left alone — it's an explicit user action and should still create a fresh one.

## Test plan

- [x] `pnpm --filter @composio/ao-cli test`
- [x] `pnpm --filter @composio/ao-core test`
- [x] `pnpm --filter @composio/ao-web test`
- [x] `pnpm --filter @composio/ao-cli typecheck`
- [x] `pnpm --filter @composio/ao-core typecheck`
- [x] `pnpm --filter @composio/ao-web typecheck`
- [x] `pnpm eslint` on changed files
- [ ] Manual smoke test: run `ao start`, observe `int-orchestrator-1` is created. Run `ao stop`, then `ao start` again — verify the same `-1` id is reused (or restored if its tmux pane died) instead of jumping to `-2`. Verify the dashboard's "orchestrator" link matches the CLI-printed id.
- [ ] Manual smoke test: kill the orchestrator's tmux pane manually, then `ao start` — verify it logs `Restoring orchestrator session: int-orchestrator-1` instead of allocating `-2`.

### CODE REVIEW

File: packages/cli/src/commands/start.ts [hunk 1]
```diff
@@ -1096,42 +1138,49 @@ async function runStartup(
...
+        `${target}${restoreNote}${otherSummarySuffix}`,
       );
     }
-  } else if (reused) {
-    console.log(chalk.cyan("Orchestrator:"), `reused existing session (${sessionId})`);
+  } else if (hasMultipleReusable) {
+    console.log(
+      chalk.cyan("Orchestrator:"),
+      "multiple sessions found — select one in the dashboard",
+    );
```
[the remark this answers is not in the document]
Comment (harsh-batheja) [reply packages/cli/src/commands/start.ts]: Already addressed in 64b3913a — the dead else-if branch was removed in the same commit that added coverage tests for the surrounding paths. Thanks for the catch.

File: packages/cli/src/commands/start.ts [hunk 2]
```diff
@@ -1433,22 +1477,53 @@ export function registerStop(program: Command): void {
...
+            await sm.kill(orchestratorToKill.id, { purgeOpenCode });
+            spinner.succeed(`Orchestrator session stopped (${orchestratorToKill.id})`);
+            // Also log to console.log so the killed id is visible in non-TTY callers
+            // (CI, scripts) and in test capture, since spinner output is suppressed.
+            console.log(chalk.green(`  Stopped orchestrator session: ${orchestratorToKill.id}`));
           } else {
-            console.log(chalk.yellow(`Orchestrator session "${sessionId}" is not running`));
+            console.log(
+              chalk.yellow(`No running orchestrator session found for "${project.name}"`),
+            );
```
[the remark this answers is not in the document]
Comment (harsh-batheja) [reply packages/cli/src/commands/start.ts:1694]: Fixed in 46de2a03 — added a `lookupFailed` flag and skip the `else` branch when sm.list threw, so only the truthful catch-block warning is shown. Updated the corresponding test to assert the misleading message is no longer printed.

File: packages/cli/src/commands/start.ts [hunk 3]
```diff
@@ -1032,33 +1035,72 @@ async function runStartup(
...
+    // an existing one is still usable:
+    //   - live:       runtime is still running, attach in place.
+    //   - restorable: status is terminal but the session can be restarted via
+    //                 sm.restore() (workspace + branch + handle still on disk).
+    //                 Restoring keeps the original numbered id rather than
+    //                 allocating a fresh one.
+    const live = orchestrators.filter((s) => !isTerminalSession(s));
+    const restorable = orchestrators.filter(
+      (s) => isTerminalSession(s) && isRestorable(s),
+    );
```
[the remark this answers is not in the document]
Comment (harsh-batheja) [reply packages/cli/src/commands/start.ts]: Fixed in 46de2a03 — dropped the redundant outer `isTerminalSession(s)` check from the `restorable` filter since `isRestorable` already requires it.

File: packages/cli/src/commands/start.ts [hunk 4]
```diff
@@ -1032,33 +1035,72 @@ async function runStartup(
...
+    const live = orchestrators.filter((s) => !isTerminalSession(s));
+    // isRestorable already requires isTerminalSession internally, so no need
+    // to repeat that guard here.
+    const restorable = orchestrators.filter((s) => isRestorable(s));
+    type OrchestratorCandidate = { session: Session; mode: "live" | "restore" };
+    const candidates: OrchestratorCandidate[] = [
+      ...live.map<OrchestratorCandidate>((session) => ({ session, mode: "live" })),
+      ...restorable.map<OrchestratorCandidate>((session) => ({ session, mode: "restore" })),
+    ];
+    candidates.sort(
```
Comment (illegalcall) [inline packages/cli/src/commands/start.ts]: candidates mixes live + restorable sessions and sorts only by lastActivityAt. 
If a live orchestrator exists but a killed one is newer, ao start restores the killed one instead of reusing the already-live session. That leaves both orchestrators alive after start. 
Repro is the exact path covered by the new test: 
live app-orchestrator-2, restorable app-orchestrator-3 newer -> sm.restore("app-orchestrator-3"), 
while -2 keeps running.

File: packages/core/src/types.ts [hunk 5]
```diff
@@ -194,7 +194,10 @@ export function isOrchestratorSession(
   sessionPrefix?: string,
   allSessionPrefixes?: string[],
 ): boolean {
-  if (session.metadata?.["role"] === "orchestrator" || session.id.endsWith("-orchestrator")) {
+  // Explicit role metadata is always authoritative — covers legacy
```
Comment (illegalcall) [inline packages/core/src/types.ts:337]: This does not affect the real `sm.list()` path. 
`sessionManager.list()` repairs bare `*-orchestrator` records first and stamps `role: "orchestrator"` in `repairSingleSessionMetadataOnRead()` (`packages/core/src/session-manager.ts`), and both `ao start` and the dashboard consume `sm.list()`. 

So these stale bare ids still pass `isOrchestratorSession()` after repair. The new tests only cover synthetic `Session` objects, not actual session-manager output.

**Review by illegalcall — APPROVED** [review]

File: packages/cli/src/commands/start.ts (hunk 4 above)
Comment (harsh-batheja) [reply packages/cli/src/commands/start.ts]: Fixed in dd0f6204. The cross-bucket sort is gone — multi-orchestrator support was removed entirely. Each project now has exactly one orchestrator at the fixed id `{sessionPrefix}-orchestrator` (no `-N`). `sm.spawnOrchestrator` is idempotent: if a record exists at the canonical id, it returns the running session or restores it in place; `runStartup` just calls it unconditionally. No partition, no sort, no choice between candidates → the bug you described is architecturally impossible now.

File: packages/core/src/types.ts (hunk 5 above)
Comment (harsh-batheja) [reply packages/core/src/types.ts:337]: Fixed in dd0f6204. Split the SM-internal predicate into `isCanonicalOrchestratorId` (strict, id-only) and `isOrchestratorSessionRecord` (inclusive, role OR id). `repairSingleSessionMetadataOnRead` and `repairSessionMetadataOnRead` both switched to the strict variant, so legacy `{prefix}-orchestrator-N` records no longer get `role: orchestrator` backfilled on read. `isOrchestratorSession` (public) is now a strict exact-id match when a prefix is given, so even real `sm.list()` output from the repaired records no longer leaks the legacy shape into the dashboard/CLI.

File: packages/cli/src/commands/start.ts (hunk 4 above)
Comment (harsh-batheja) [reply packages/cli/src/commands/start.ts]: Addressed in-place in 5dafe1b1 (multi-orchestrator support preserved). `runStartup` now prefers the `live` bucket unconditionally — it only drops to `restorable` when zero live candidates exist. A live-older vs killed-newer race can no longer pick the killed one, so `sm.restore()` can't resurrect a record while another is still running. New regression test "prefers a live orchestrator over a more-recently-active restorable one" pins the priority.

File: packages/core/src/types.ts (hunk 5 above)
Comment (harsh-batheja) [reply packages/core/src/types.ts:337]: Addressed in-place in 5dafe1b1 (multi-orchestrator support preserved). Added a stricter sibling predicate `isRepairableOrchestratorRecord` used only by the repair-on-read path. It accepts role-stamped records, the bare `{sessionPrefix}-orchestrator` shape (correct-prefix legacy), and `{sessionPrefix}-orchestrator-N`, but rejects foreign bare names like `{projectId}-orchestrator`. So legacy `-` shapes with wrong prefix no longer get `role: orchestrator` backfilled on read and therefore no longer pass `isOrchestratorSession()` in real `sm.list()` output. `isOrchestratorSessionRecord` stays permissive for cleanup protection. New regression test in query.test.ts writes a `my-app-orchestrator` record under sessionPrefix `app`, runs `sm.list()`, and asserts `role` is still absent on disk afterward.

**Comment by whoisasx** [comment]
## Review Summary

Reviewed the full diff. This PR correctly addresses the core problem — `ao start` allocating a fresh `-orchestrator-N` on every invocation instead of reusing/restoring the previous one.

### What it does

**Three coordinated fixes:**

1. **Reuse-or-restore in `runStartup`** — Partitions existing orchestrators into `live` (still running) and `restorable` (tmux pane died but session data intact) buckets. Live is preferred unconditionally over restorable to avoid spinning up two orchestrators. Only spawns fresh when both buckets are empty.

2. **Eliminates the phantom `{prefix}-orchestrator` id** — All URL prints, browser-open targets, and `ao stop` kill targets now use the real numbered id resolved via `sm.list()`. Previously `ao stop` was calling `sm.kill("${prefix}-orchestrator")` against an id that never matched a real numbered session.

3. **Tightens `isOrchestratorSession` in `types.ts`** — Removes the `endsWith("-orchestrator")` fallback. Bare legacy records now require explicit `role: orchestrator` metadata (backfilled by `repairSingleSessionMetadataOnRead`). Adds `isRepairableOrchestratorRecord` in session-manager to prevent stamping `role` on foreign bare-id records.

### What looks good

- The live-over-restorable priority is correct and well-documented — avoids the dual-orchestrator race the comment explains.
- `ao stop` using `sm.list()` lookup instead of the hardcoded phantom id is the right fix.
- Restore failure is surfaced clearly (spinner.fail + thrown error with cause chain) rather than silently falling through to spawn.
- Test coverage is thorough: restore-on-killed, stale-bare exclusion, cross-bucket recency, live-multi reuse, restore failure cleanup, and `--no-orchestrator` fallback URL.

### Minor observations

- `isRepairableOrchestratorRecord` duplicates some logic from `isOrchestratorSessionRecord` — could share a base, but keeping them separate is reasonable given they serve different purposes (repair eligibility vs. general detection).
- The `ao stop` path now does a full `sm.list()` + filter + sort just to find one id to kill. Fine for correctness, but worth noting the extra I/O vs. the old direct `sm.get()`.

### Verdict

Looks good to merge. The fix is well-scoped, the edge cases are covered, and the test suite pins down the new contract.

**Review by whoisasx — APPROVED** [review]

File: packages/core/src/session-manager.ts [hunk 6]
```diff
@@ -401,7 +429,10 @@ export function createSessionManager(deps: SessionManagerDeps): OpenCodeSessionM
     sessionPrefix?: string,
   ): ActiveSessionRecord {
     const repaired = { ...record, raw: { ...record.raw } };
-    if (!isOrchestratorSessionRecord(repaired.sessionName, repaired.raw, sessionPrefix)) {
+    // Use the strict repairable predicate: bare legacy `*-orchestrator`
+    // records are not eligible for role backfill, so they cannot leak into
+    // `isOrchestratorSession` via a stamped role on the next sm.list().
```
[the remark this answers is not in the document]
Comment (harsh-batheja) [reply packages/core/src/session-manager.ts]: Fixed in a8bf0424. Used your suggested wording — the comment now clarifies that only *foreign* bare legacy records (wrong prefix) are excluded from repair/backfill; correct-prefix bare `{sessionPrefix}-orchestrator` records are still repaired. Thanks for the catch.

File: packages/cli/src/commands/start.ts [hunk 7]
```diff
@@ -1060,33 +1063,80 @@ async function runStartup(
...
+    //                 sm.restore() (workspace + branch + handle still on disk).
+    //                 Restoring keeps the original numbered id rather than
+    //                 allocating a fresh one.
+    //
+    // IMPORTANT: live MUST be preferred unconditionally over restorable. A
+    // previous version sorted both buckets together by `lastActivityAt`, which
+    // could pick a newer killed record over an older-but-still-running one —
+    // sm.restore() would then spin up the killed record while the live one
+    // kept running, leaving two orchestrators alive for the project. Only fall
+    // back to restorable when the live bucket is empty.
```
[the remark this answers is not in the document]
Comment (harsh-batheja) [reply packages/cli/src/commands/start.ts:1128]: Updated the PR description to match the implementation. The live-over-restorable priority is intentional — an earlier version did sort across both buckets and that was flagged by @illegalcall (comment above) as a bug: a newer killed record could beat an older-but-still-running one, so `sm.restore()` would resurrect the killed record while the live one kept running. Preferring live unconditionally prevents that. PR body now states 'Live is preferred unconditionally over restorable' explicitly, and the code comment at start.ts:1083 spells out the reasoning.

**Comment by ashish921998** [comment]
@harsh-batheja Please resolve conflicts here

**Comment by harsh-batheja** [comment]
Follow-up filed: #1323 — the in-memory `lifecycle.session.kind` leak for foreign-prefix legacy records that the rebased canonical-lifecycle code exposed but this PR only partially closes. Deliberately kept out of scope for #1075 since the trigger (unusual legacy disk state) is rare and the fix is self-contained.
